File: 이것이 코딩테스트다/06.다이나믹 프로그래밍/6.효율적인 화폐 구성.py
n, m = map(int, input().split())
moneys = []
for _ in range(n):
  moneys.append(int(input()))

INF = int(1e9)
dp = [INF] * (m+1)
min_money = min(moneys)

for money in moneys:
  if m >= money:
    dp[money] = 1

for i in range(min_money+1, m+1):
  for money in moneys:
    if i-money > 0:
      dp[i] = min(dp[i], dp[i-money]+1)

if dp[m] == INF:
  print(-1)
else:
  print(dp[m])

File: 이것이 코딩테스트다/06.다이나믹 프로그래밍/6.효율적인 화폐 구성.js
const texts = [2, 15, [2, 3]];

const n = texts.shift();
const m = texts.shift();

const money = texts.pop();
const INF = Number(Infinity);
let dp = Array(m + 1).fill(INF);
const minMoney = Math.min(...money);
for (const m of money) {
  dp[m] = 1;
}
for (let i = 1; i < m + 1; i++) {
  for (const m of money) {
    if (i - m >= 0 && dp[i - m] !== INF) {
      dp[i] = Math.min(dp[i], dp[i - m] + 1);
    }
  }
}

if (dp[m] === INF) {
  console.log(-1);
} else {
  console.log(dp[m]);
}
